db-tests/src/test/java/TestTriggers.java

```diff
@@ -577,4 +639,124 @@ void shouldIncrementSimulationRevisionOnSimulationUpdate() throws SQLException {
       assertEquals(initialRevision + 1, updatedRevision);
     }
   }
+
+  @Nested
+  class PlanDatasetTriggers {
+    @Test
+    void shouldCreateDefaultDatasetOnPlanDatasetInsertWithNullDatasetId() throws SQLException {
+      final var res = connection.createStatement()
+          .executeQuery(
+              """
+                  INSERT INTO plan_dataset (plan_id, offset_from_plan_start)
+                  VALUES (%s, '0')
+                  RETURNING dataset_id;"""
+                  .formatted(planId)
+          );
+      res.next();
+      final var newDatasetId = res.getInt("dataset_id");
+      res.close();
+
+      assertInstanceOf(Integer.class, newDatasetId);
+
+      final var datasetRes = connection.createStatement()
+          .executeQuery(
+              """
+                  SELECT * FROM dataset
+                  WHERE id = %s;"""
+                  .formatted(newDatasetId)
+          );
+
+      datasetRes.next();
+      assertEquals(newDatasetId, datasetRes.getInt("id"));
+      assertEquals(0, datasetRes.getInt("revision"));
+      datasetRes.close();
+    }
+
+    @Test
+    void shouldCalculatePlanDatasetOffsetOnPlanDatasetInsertWithNonNullDatasetId() throws SQLException {
+
+      final var planRes = connection.createStatement()
+          .executeQuery(
+              """
+                  SELECT * from plan
+                  WHERE id = %s;"""
+                  .formatted(planDatasetId.getLeft())
+          );
+      planRes.next();
+      final var planStartTime = planRes.getTimestamp("start_time");
+      planRes.close();
+
+      final var planDatasetSelectRes = connection.createStatement()
+          .executeQuery(
+              """
+                  SELECT * FROM plan_dataset
+                  WHERE plan_id = %s and dataset_id = %s;"""
+                  .formatted(planDatasetId.getLeft(), planDatasetId.getRight())
+          );
+      planDatasetSelectRes.next();
+      final var offsetFromPlanStart = Duration.parse(planDatasetSelectRes.getString("offset_from_plan_start"));
+      planDatasetSelectRes.close();
+      assertEquals(Duration.ofMillis(0), offsetFromPlanStart);
+
+      final var newPlanId = insertPlan(missionModelId, "2020-1-1 01:00:00");
+
+      final var newPlanRes = connection.createStatement()
+          .executeQuery(
+              """
+                  SELECT * from plan
+                  WHERE id = %s;"""
+                  .formatted(newPlanId)
+          );
+      newPlanRes.next();
+      final var newPlanStartTime = newPlanRes.getTimestamp("start_time");
+      newPlanRes.close();
```

Comment:
All of this is setting up initial conditions for the actual test which is that an inserted plan_dataset with a dataset_id should automatically compute the offset_from_plan_start.

It would be better to have the initial state before this state have two plans and one plan_dataset already in the database, and have the required information (start times, and offset) stored in variables so you don't have to read the database to get them. You're already doing this anyway with a bit of other information, it won't hurt to add some more initial state.

Reply:
I wish we could have a `@Before` annotation that applies only to this specific test... I checked before I wrote my above comment and it's not included in the API we are using, so that's not an option.

Maybe you could just move the setup to a separate function just above this and call it with a comment explaining that the function just sets up initial state for the test? Then if a failure happens in the initialization, at least it's clear from the stack trace it's in the initialization function, and not the test itself.

Reply:
Well, since there's no elegant solution here I'm fine with keeping it how it is, so I'll resolve.